### ISSUE TITLE
cmake complains about "GateLib"

### ISSUE THREAD

**Issue by djboersma** [OpenGATE/Gate#320]
On Ubuntu 18.04 (with cmake version 3.10.2) I cannot build Gate, because cmake runs into this error:

```
CMake Error at CMakeLists.txt:283 (TARGET_LINK_LIBRARIES):
  Object library target "GateLib" may not link to anything.
```

On my personal laptop with a newer OS and cmake 3.13.4 it runs fine, so maybe this error is just a quirk of the cmake version 3.10.2 that happens to be installed on Ubuntu 18.04. But Gate should really be buildable on a vanilla Ubuntu 18.04 system. So either we make it work with 3.10.2 or (less preferred) we explicitly require a "new enough" cmake with the `CMAKE_MINIMUM_REQUIRED` command, using a much more recent version than we do now (3.0).

**Comment by tbaudier** [comment]
I tried to debug but I am not an expert of Cmake.
It seems that Object Librairies (like GateLib) do not have to be linked to anything:
https://stackoverflow.com/questions/53416527/cmake-object-library-target-may-not-link-to-anything

But without the line https://github.com/OpenGATE/Gate/blob/develop/CMakeLists.txt#L283 , it fails to build (see https://github.com/OpenGATE/Gate/pull/318 following https://github.com/OpenGATE/Gate/pull/290)

**Comment by djboersma** [comment]
I really hope that we can make it work for the cmake version that is installed on vanilla Ubuntu 18.04 systems, because that (and even older systems) is what many users use. But simply setting the minimal cmake version to 3.12 seems to be *much* less work for us...
I think that on Ubuntu 18.04, there is an option to install the newest cmake with "snap" instead of with "apt". That might be a solution for some users, but not for all. Some sysadmins are (understandably) quite conservative.

BTW: the `GateLib` object library is now defined in the `ELSE` leg of the `IF(GATE_USE_GPU AND CUDA_FOUND)`, but used outside of that if-else construct. This won't work if anyone tries to build with CUDA. Maybe this is the moment to officially and practically remove the CUDA code, after deprecating its use for more than two years?

**Comment by tbaudier** [comment]
> I really hope that we can make it work for the cmake version that is installed on vanilla Ubuntu 18.04 systems, because that (and even older systems) is what many users use. But simply setting the minimal cmake version to 3.12 seems to be _much_ less work for us...

I agree with you, set the minimal cmake version is easier. And I do not understand everything about the concept of Object Librairies. Maybe @wrzof can debug it for older CMake version?

> I think that on Ubuntu 18.04, there is an option to install the newest cmake with "snap" instead of with "apt". That might be a solution for some users, but not for all. Some sysadmins are (understandably) quite conservative.

On my computer, I do not use the package manager to install CMake. I just download the Linux binary distribution from https://cmake.org/download/ and set the bin path in PATH variable. It works without admin rights.

> 
> BTW: the `GateLib` object library is now defined in the `ELSE` leg of the `IF(GATE_USE_GPU AND CUDA_FOUND)`, but used outside of that if-else construct. This won't work if anyone tries to build with CUDA. Maybe this is the moment to officially and practically remove the CUDA code, after deprecating its use for more than two years?

This is something I noticed without understand it. And it's the same case for Gate and other binaries, if I understand well, isn't it?

**Comment by dsarrut** [comment]
should be solved with #322
